### PR TITLE
Making credential providers async.

### PR DESCRIPTION
**Pull request by mtdowling**
- Moving serialize from client to functions.php.
- Removed refreshable credentials. This is something that will be done
  in the memoized credential provider.
- Updated credential collaborators to use promise->wait()

### CODE REVIEW

File: src/S3/Transfer.php [hunk 1]
```diff
@@ -338,6 +341,7 @@ private function addDebugToBefore($debug)
                     $source = "s3://{$command['Bucket']}/{$command['Key']}";
                     $dest = $command['@http']['sink'];
                     break;
+                case 'CompleteMultipartUpload':
```
Comment (mtdowling) [inline src/S3/Transfer.php]: This was missing from the debug list for some reason

File: src/S3/Transfer.php [hunk 2]
```diff
@@ -277,7 +276,12 @@ private function createUploadPromise()
 
         // Create an EachPromise, that will concurrently handle the upload
         // operations' yielded promises from the iterator.
-        return $this->promise = Promise\each_limit($files, $this->concurrency);
+        return Promise\each_limit(
+            $files,
+            $this->concurrency,
+            null,
+            $this->getRejectFn()
```
Comment (mtdowling) [inline src/S3/Transfer.php]: This ensures that the first failed transfer fails the aggregate promise

File: src/S3/Transfer.php [hunk 3]
```diff
@@ -254,7 +252,8 @@ private function createDownloadPromise()
             // Create a GetObject command pool and return the promise.
             return (new Aws\CommandPool($this->client, $commands, [
                 'concurrency' => $this->concurrency,
-                'before'      => $this->before
+                'before'      => $this->before,
+                'rejected'    => $this->getRejectFn()
```
Comment (mtdowling) [inline src/S3/Transfer.php]: This ensures that the first failed transfer fails the aggregate promise.

**Comment by jeremeamia** [comment]
Reviewing this now.

File: src/S3/Transfer.php (hunk 1 above)
Comment (jeremeamia) [inline src/S3/Transfer.php]: It's already there, a few lines down.

**Comment by jeremeamia** [comment]
Looks good, except for the one comment I made on the S3 Transfer changes.

File: src/S3/Transfer.php (hunk 1 above)
Comment (mtdowling) [inline src/S3/Transfer.php]: Nah, all that's there is:

- PutObject
- UploadPart
- CreateMultipartUpload
- CompleteUploadPart

Should I just replace `CompleteUploadPart` with `CompleteMultipartUpload`?

Comment (jeremeamia) [inline src/S3/Transfer.php]: Yep. Apparently, I can't read or write.

**Comment by mtdowling** [comment]
I've fixed the CompleteMultipartUpload issue with the switch statement.

**Pull request merged**